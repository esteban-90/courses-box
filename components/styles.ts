import { css, SerializedStyles } from '@emotion/react'
import { AppTheme } from '@/styles/themes'

export type Color = 'primary' | 'secondary' | 'danger' | 'warning'

export const boxShadow = (shadowColor1: string, shadowColor2: string, inset = false): SerializedStyles => {
  const insetStr = inset ? 'inset' : false

  return css`
    box-shadow: 0.5vmin 0.5vmin 1vmin ${shadowColor1} ${insetStr}, -0.5vmin -0.5vmin 1vmin ${shadowColor2} ${insetStr};
  `
}

export const transition = (): SerializedStyles => {
  return css`
    transition: all 0.4s ease;
  `
}

export const getColors = (theme: AppTheme, color?: Color): SerializedStyles => {
  switch (color) {
    case 'secondary':
      return css`
        color: ${theme.font.regular};
      `

    case 'primary':
    case 'danger':
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.button};
      `

    case 'warning':
      return css`
        background: ${theme.components[color]};
        color: ${theme.font.warning};
      `

    default:
      return css``
  }
}
