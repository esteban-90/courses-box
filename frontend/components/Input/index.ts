export * from './Input'
export * from './Feedback'
