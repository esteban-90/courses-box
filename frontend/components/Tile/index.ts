export * from './Tile'
export * from './CenteredTile'
