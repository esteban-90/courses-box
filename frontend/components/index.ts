export * from './Button'
export * from './Checkbox'
export * from './Course'
export * from './Icon'
export * from './IconButton'
export * from './Input'
export * from './Layout'
export * from './Link'
export * from './Logo'
export * from './Switch'
export * from './Tile'
