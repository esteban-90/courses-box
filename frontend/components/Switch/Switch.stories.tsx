import type { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { screen, userEvent } from '@storybook/testing-library'
import { Switch } from '@/components/Switch'

export default {
  title: 'Controls/Switch',
  component: Switch,
} as ComponentMeta<typeof Switch>

export const BasicSwitch: ComponentStoryObj<typeof Switch> = {
  async play() {
    await userEvent.click(screen.getByTestId('switchVisiblePart'))
  },
}
