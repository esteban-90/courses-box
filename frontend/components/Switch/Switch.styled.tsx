import styled from '@emotion/styled'
import { makeShadow } from '@/styles/helpers'

export const VisiblePart = styled.label`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
  height: 3rem;
  width: 6rem;
  border-radius: 1.6rem;
  background-color: ${({ theme }) => theme.components.background};

  box-shadow: ${({ theme }) => {
    const { shadow1, shadow2 } = theme.components
    const $shadow = makeShadow(shadow1, shadow2)
    return $shadow
  }};

  &::after {
    content: '';
    margin-left: 0.5rem;
    width: 2.1rem;
    height: 2.1rem;
    border-radius: 50%;
    background-color: ${({ theme }) => theme.components.nonActive};
    transition: all 0.4s ease;
  }
`

export const Wrapper = styled.label`
  & > input {
    display: none;
  }

  & > input:checked {
    & ~ label {
      background-color: ${({ theme }) => theme.components.primary};

      &::after {
        margin-left: 3.5rem;
        background-color: ${({ theme }) => theme.components.active};
      }
    }
  }
`
