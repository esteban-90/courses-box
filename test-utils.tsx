import { FC, ReactElement, ReactNode } from 'react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Themes } from '@/styles/themes'

const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }

// 14
