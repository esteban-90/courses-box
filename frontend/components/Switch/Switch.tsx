import { FC } from 'react'
import { useId } from '@/hooks'
import { SwitchProps } from '@/types'
import * as Styled from './Switch.styled'

export const Switch: FC<SwitchProps> = ({ onChange }): JSX.Element => {
  const inputId = useId()

  return (
    <Styled.Wrapper>
      <input id={inputId} type='checkbox' onChange={onChange} />
      <Styled.VisiblePart htmlFor={inputId} data-testid='SwitchVisiblePart' />
    </Styled.Wrapper>
  )
}
