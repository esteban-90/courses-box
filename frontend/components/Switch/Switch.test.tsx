import userEvent from '@testing-library/user-event'
import { render, screen } from '@/test-utils'
import { Switch } from '@/components/Switch'

describe('Checkbox test cases', () => {
  const changeHandler = jest.fn()
  const switchElement = <Switch onChange={changeHandler} />

  it('Render check', () => {
    jest.spyOn(Math, 'random').mockReturnValue(0.999_999_999)
    const { asFragment } = render(switchElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Check onChange callback', async () => {
    render(switchElement)
    await userEvent.click(screen.getByTestId('switchVisiblePart'))
    expect(changeHandler).toHaveBeenCalled()
  })
})
