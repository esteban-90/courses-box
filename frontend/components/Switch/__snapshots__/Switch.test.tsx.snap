// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox test cases Render check 1`] = `
<DocumentFragment>
  <label
    class="css-1wwxbmj"
  >
    <input
      id="prefix-fffffffbb47d08"
      type="checkbox"
    />
    <label
      class="css-1kzb2zn"
      data-testid="switchVisiblePart"
      for="prefix-fffffffbb47d08"
    />
  </label>
</DocumentFragment>
`;
